fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/query/plan/debug/Commands.java

```diff
@@ -0,0 +1,623 @@
+            return "HELP";
+        }
+
+        @Override
+        public void printUsage(@Nonnull final PlannerRepl plannerRepl) {
+            plannerRepl.printlnKeyValue("help", "print this help");
+        }
+    }
+
+    /**
+     * List out all references.
+     */
+    @AutoService(Command.class)
+    public static class RefsCommand implements Command<Event> {
+        @Override
+        public boolean executeCommand(@Nonnull final PlannerRepl plannerRepl,
+                                      @Nonnull final Event event,
+                                      @Nonnull final ParsedLine parsedLine) {
+            final State state = plannerRepl.getCurrentState();
+            final Cache<Integer, ExpressionRef<? extends RelationalExpression>> referenceCache = state.getReferenceCache();
+            final List<Integer> ids = Lists.newArrayList(referenceCache.asMap().keySet());
+            Collections.sort(ids);
+            for (Integer id : ids) {
+                plannerRepl.printKeyValue("id", "ref" + id + "; ");
+                @Nullable final ExpressionRef<? extends RelationalExpression> reference = referenceCache.getIfPresent(id);
+                plannerRepl.printKeyValue("kind", reference == null ? "not in cache; " : reference.getClass().getSimpleName() + "; ");
+                if (reference instanceof GroupExpressionRef) {
+                    final GroupExpressionRef<? extends RelationalExpression> groupReference = (GroupExpressionRef<? extends RelationalExpression>)reference;
+                    final String membersString = groupReference.getMembers()
+                            .stream()
+                            .map(expression -> Optional.ofNullable(plannerRepl.nameForObject(expression)))
+                            .filter(Optional::isPresent)
+                            .map(Optional::get)
+                            .collect(Collectors.joining(", "));
+                    plannerRepl.printKeyValue("members: ", "{" + membersString + "}");
+                }
+                plannerRepl.println();
+            }
+            return false;
+        }
+
+        @Nonnull
+        @Override
+        public String getCommandToken() {
+            return "REFS";
+        }
+
+        @Override
+        public void printUsage(@Nonnull final PlannerRepl plannerRepl) {
+            plannerRepl.printlnKeyValue("refs", "list all references");
+        }
+    }
+
+    /**
+     * Restart execution. All entity names remain stable.
+     */
+    @AutoService(Command.class)
+    public static class RestartCommand implements Command<Event> {
+        @Override
+        public boolean executeCommand(@Nonnull final PlannerRepl plannerRepl,
+                                      @Nonnull final Event event,
+                                      @Nonnull final ParsedLine parsedLine) {
+            plannerRepl.restartState();
+            plannerRepl.addInternalBreakPoint(new PlannerRepl.CountingTautologyBreakPoint(1));
```

Comment:
Does this mean that restart stops after the _first_ thing the new plan does or before it?

Reply:
The first thing the compilation process does it to pop the first task from the task stack and call the debugger with an instance of `ExecutingTaskEvent`. That's when the break point stops execution again and self-removes itself from the list of break points. So to answer your question it will be before the first actual thing happens and identical to what happens when you first run it.